stage1/app-add/app-add.go

```diff
@@ -165,10 +201,389 @@ func prepareAppCgroups(p *stage1types.Pod, ra *schema.RuntimeApp, enterCmd []str
+		log.Printf,
+	)
+
+	// This logic is mostly copied from appToNspawnArgs
+	// TODO(cdc): deduplicate
+	for _, m := range mounts {
+		shPath := filepath.Join(sharedVolPath, m.Volume.Name.String())
+
+		// Evaluate symlinks within the app's rootfs - otherwise absolute
+		// symlinks will be wrong.
+		mntPath, err := stage1init.EvaluateSymlinksInsideApp(appRootfs, m.Mount.Path)
+		if err != nil {
+			return errwrap.Wrap(fmt.Errorf("Could not evaluate path %v", m.Mount.Path), err)
+		}
+		mntAbsPath := filepath.Join(appRootfs, mntPath)
+
+		// Create the stage1 destination
+		if err := stage1init.PrepareMountpoints(shPath, mntAbsPath, &m.Volume, m.DockerImplicit); err != nil {
+			return errwrap.Wrapf("could not prepare mountpoint", err)
+		}
+
+		err = addMountStage0(p, ra, &m, mnt, enterCmd)
+		if err != nil {
+			return errwrap.Wrap(fmt.Errorf("error adding mount volume %v path %v", m.Mount.Volume, m.Mount.Path), err)
+		}
+	}
+	return nil
+}
+
+// addMountStage1 bind-mounts (moves) the given mount from the host in the container.
+//
+// We use the propagation mechanism of systemd-nspawn. In all systemd-nspawn
+// containers, the directory "/run/systemd/nspawn/propagate/$MACHINE_ID" on
+// the host is propagating mounts to the directory
+// "/run/systemd/nspawn/incoming/" in the container mount namespace. Once a
+// bind mount is propagated, we simply move to its correct location.
+//
+// The algorithm is the same as in "machinectl bind":
+// https://github.com/systemd/systemd/blob/v231/src/machine/machine-dbus.c#L865
+// except that we don't use setns() to enter the mount namespace of the pod
+// because Linux does not allow multithreaded applications (such as Go
+// programs) to change mount namespaces with setns. Instead, we fork another
+// process written in C (single-threaded) to enter the mount namespace. The
+// command used is specified by the "enterCmd" parameter.
+//
+// Users might request a bind mount to be set up read-only. This complicates
+// things a bit because on Linux, setting up a read-only bind mount involves
+// two mount() calls, so it is not atomic. We don't want the container to see
+// the mount in read-write mode, even for a short time, so we don't create the
+// bind mount directly in "/run/systemd/nspawn/propagate/$MACHINE_ID" to avoid
+// an immediate propagation to the container. Instead, we create a temporary
+// playground in "/tmp/rkt.propagate.XXXX" and create the bind mount in
+// "/tmp/rkt.propagate.XXXX/mount" with the correct read-only attribute before
+// moving it.
+//
+// Another complication is that the playground cannot be on a shared mount
+// because Linux does not allow MS_MOVE to be applied to mounts with MS_SHARED
+// parent mounts. But by default, systemd mounts everything as shared, see:
+// https://github.com/systemd/systemd/blob/v231/src/core/mount-setup.c#L392
+// We set up the temporary playground as a slave bind mount to avoid this
+// limitation.
+func addMountStage0(
```

Comment:
What's the reasoning behind moving this out of `common`? It really seems like library code, not `main` code.